docs/manpages/vfs_audit.8
'\" t
.\"     Title: vfs_audit
.\"    Author: [see the "AUTHOR" section]
.\" Generator: DocBook XSL Stylesheets v1.79.1 <http://docbook.sf.net/>
.\"      Date: 07/02/2020
.\"    Manual: System Administration tools
.\"    Source: Samba 4.12.5
.\"  Language: English
.\"
.TH "VFS_AUDIT" "8" "07/02/2020" "Samba 4\&.12\&.5" "System Administration tools"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
vfs_audit \- record selected Samba VFS operations in the system log
.SH "SYNOPSIS"
.HP \w'\ 'u
vfs objects = audit
.SH "DESCRIPTION"
.PP
This VFS module is part of the
\fBsamba\fR(7)
suite\&.
.PP
The
vfs_audit
VFS module records selected client operations to the system log using
\fBsyslog\fR(3)\&.
.PP
The following Samba VFS operations are recorded:
.RS 4
connect
.RE
.RS 4
disconnect
.RE
.RS 4
opendir
.RE
.RS 4
mkdir
.RE
.RS 4
rmdir
.RE
.RS 4
open
.RE
.RS 4
close
.RE
.RS 4
rename
.RE
.RS 4
unlink
.RE
.RS 4
chmod
.RE
.RS 4
fchmod
.RE
.PP
This module is stackable\&.
.SH "OPTIONS"
.PP
audit:facility = FACILITY
.RS 4
Log messages to the named
\fBsyslog\fR(3)
facility\&.
.RE
.PP
audit:priority = PRIORITY
.RS 4
Log messages with the named
\fBsyslog\fR(3)
priority\&.
.RE
.SH "EXAMPLES"
.PP
Log operations on all shares using the LOCAL1 facility and NOTICE priority:
.sp
.if n \{\
.RS 4
.\}
.nf
        \fI[global]\fR
	\m[blue]\fBvfs objects = audit\fR\m[]
	\m[blue]\fBaudit:facility = LOCAL1\fR\m[]
	\m[blue]\fBaudit:priority = NOTICE\fR\m[]
.fi
.if n \{\
.RE
.\}
.SH "VERSION"
.PP
This man page is part of version 4\&.12\&.5 of the Samba suite\&.
.SH "AUTHOR"
.PP
The original Samba software and related utilities were created by Andrew Tridgell\&. Samba is now developed by the Samba Team as an Open Source project similar to the way the Linux kernel is developed\&.


debian/NEWS
samba (2:4.3.7+dfsg-1) unstable; urgency=high

    This Samba security addresses both Denial of Service and Man in
    the Middle vulnerabilities.

    Both of these changes implement new smb.conf options and a number
    of stricter behaviours to prevent Man in the Middle attacks on our
    network services, as a client and as a server.

    Between these changes, compatibility with a large number of older
    software versions has been lost in the default configuration.

    See the release notes in WHATNEW.txt for more information.


    Here are some additional hints how to work around the new stricter default behaviors:

    * As an AD DC server, only Windows 2000 and Samba 3.6 and above as
      a domain member are supported out of the box. Other smb file
      servers as domain members are also fine out of the box.

    * As an AD DC server, with default setting of "ldap server require
      strong auth", LDAP clients connecting over ldaps:// or START_TLS
      will be allowed to perform simple LDAP bind only.

      The preferred configuration for LDAP clients is to use SASL
      GSSAPI directly over ldap:// without using ldaps:// or
      START_TLS.

      To use LDAP with START_TLS and SASL GSSAPI (either Kerberos or
      NTLMSSP) sign/seal protection must be used by the client and
      server should be configured with "ldap server require strong
      auth = allow_sasl_over_tls".

      Consult OpenLDAP documentation how to set sign/seal protection
      in ldap.conf.

      For SSSD client configured with "id_provider = ad" or
      "id_provider = ldap" with "auth_provider = krb5", see
      sssd-ldap(5) manual for details on TLS session handling.

    * As a File Server, compatibility with the Linux Kernel cifs
      client depends on which configuration options are selected, please
      use "sec=krb5(i)" or "sec=ntlmssp(i)", not "sec=ntlmv2".

    * As a file or printer client and as a domain member, out of the
      box compatibility with Samba less than 4.0 and other SMB/CIFS
      servers, depends on support for SMB signing or SMB2 on the
      server, which is often disabled or absent. You may need to
      adjust the "client ipc signing" to "no" in these cases.

    * In case of an upgrade from versions before 4.2.0, you might run
      into problems as a domain member. The out of the box compatibility
      with Samba 3.x domain controllers requires NETLOGON features only
      available in Samba 3.2 and above.
    
    However, all of these can be worked around by setting smb.conf
    options in Samba, see WHATSNEW.txt the 4.2.0 release notes at
    https://www.samba.org/samba/history/samba-4.2.0.html and the Samba
    wiki for details, workarounds and suggested security-improving
    changes to these and other software packages.


    Suggested further improvements after patching:

    It is recommended that administrators set these additional options,
    if compatible with their network environment:

        server signing = mandatory
        ntlm auth = no

    Without "server signing = mandatory", Man in the Middle attacks
    are still possible against our file server and
    classic/NT4-like/Samba3 Domain controller. (It is now enforced on
    Samba's AD DC.) Note that this has heavy impact on the file server
    performance, so you need to decide between performance and
    security. These Man in the Middle attacks for smb file servers are
    well known for decades.

    Without "ntlm auth = no", there may still be clients not using
    NTLMv2, and these observed passwords may be brute-forced easily using
    cloud-computing resources or rainbow tables.

 -- Andrew Bartlett <abartlet+debian@catalyst.net.nz>  Tue, 12 Apr 2016 16:18:57 +1200

samba (2:4.0.10+dfsg-3) unstable; urgency=low

    The SWAT package is no longer available.

    Upstream support for SWAT (Samba Web Administration Tool) was removed in
    samba 4.1.0. As a result, swat is no longer shipped in the Debian Samba
    packages. Unfortunately, there is currently no replacement.

    Details why SWAT has been removed upstream can be found on the
    samba-technical mailing list:

    https://lists.samba.org/archive/samba-technical/2013-February/090572.html

 -- Ivo De Decker <ivo.dedecker@ugent.be>  Tue, 22 Oct 2013 07:52:54 +0200

samba (2:3.6.5-2) unstable; urgency=low

    NSS modules have been split out from libpam-winbind to
    libnss-winbind.
    
    If Recommends: installs are disabled on your system you may need
    to manually install the libnss-winbind package after upgrading
    from former versions of winbind (for instance from squeeze) or
    from former versions of libpam-winbind.

 -- Christian Perrier <bubulle@debian.org>  Mon, 07 May 2012 22:16:32 +0200

samba (2:3.5.11~dfsg-3) unstable; urgency=low

    PAM modules and NSS modules have been split out from the winbind
    package into libpam-winbind.
    
    If Recommends: installs are disabled on your system you may need
    to manually install the libpam-winbind package after upgrading
    from former versions of winbind (for instance from squeeze)

 -- Steve Langasek <vorlon@debian.org>  Fri, 21 Oct 2011 20:00:13 +0000

samba (2:3.4.0-1) unstable; urgency=low

    Default passdb backend changed in samba 3.4.0 and above
  
    Beginning with samba 3.4.0, the default setting for "passdb
    backend" changed from "smbpasswd" to "tdbsam".
    
    If your smb.conf file does not have an explicit mention of
    "passdb backend" when upgrading from pre-3.4.0 versions of
    samba, it is likely that users will no longer be able to
    authenticate.
    
    As a consequence of all this, if you're upgrading from lenny
    and have no setting of "passdb backend" in smb.conf, you MUST
    add "passdb backend = smbpasswd" in order to keep your samba
    server's behaviour.
    
    As Debian packages of samba explicitly set "passdb backend = tdbsam"
    by default since etch, very few users should need to modify their

source3/modules/vfs_audit.c
/* 
 * Auditing VFS module for samba.  Log selected file operations to syslog
 * facility.
 *
 * Copyright (C) Tim Potter			1999-2000
 * Copyright (C) Alexander Bokovoy		2002
 * Copyright (C) Stefan (metze) Metzmacher	2002
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *  
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *  
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, see <http://www.gnu.org/licenses/>.
 */


#include "includes.h"
#include "system/filesys.h"
#include "system/syslog.h"
#include "smbd/smbd.h"
#include "lib/param/loadparm.h"

#undef DBGC_CLASS
#define DBGC_CLASS DBGC_VFS

static int audit_syslog_facility(vfs_handle_struct *handle)
{
	static const struct enum_list enum_log_facilities[] = {
		{ LOG_AUTH,	"AUTH" },
		{ LOG_CRON,	"CRON" },
		{ LOG_DAEMON,	"DAEMON" },
		{ LOG_FTP,	"FTP" },
		{ LOG_KERN,	"KERN" },
		{ LOG_LPR,	"LPR" },
		{ LOG_MAIL,	"MAIL" },
		{ LOG_NEWS,	"NEWS" },
		{ LOG_NTP,	"NTP" },
		{ LOG_SECURITY,	"SECURITY" },
		{ LOG_SYSLOG,	"SYSLOG" },
		{ LOG_USER,	"USER" },
		{ LOG_UUCP,	"UUCP" },
		{ LOG_LOCAL0,	"LOCAL0" },
		{ LOG_LOCAL1,	"LOCAL1" },
		{ LOG_LOCAL2,	"LOCAL2" },
		{ LOG_LOCAL3,	"LOCAL3" },
		{ LOG_LOCAL4,	"LOCAL4" },
		{ LOG_LOCAL5,	"LOCAL5" },
		{ LOG_LOCAL6,	"LOCAL6" },
		{ LOG_LOCAL7,	"LOCAL7" },
		{ -1,		NULL }
	};

	int facility;

	facility = lp_parm_enum(SNUM(handle->conn), "audit", "facility", enum_log_facilities, LOG_USER);

	return facility;
}


static int audit_syslog_priority(vfs_handle_struct *handle)
{
	static const struct enum_list enum_log_priorities[] = {
		{ LOG_EMERG, "EMERG" },
		{ LOG_ALERT, "ALERT" },
		{ LOG_CRIT, "CRIT" },
		{ LOG_ERR, "ERR" },
		{ LOG_WARNING, "WARNING" },
		{ LOG_NOTICE, "NOTICE" },
		{ LOG_INFO, "INFO" },
		{ LOG_DEBUG, "DEBUG" },
		{ -1, NULL }
	};

	int priority;

	priority = lp_parm_enum(SNUM(handle->conn), "audit", "priority",
				enum_log_priorities, LOG_NOTICE);
	if (priority == -1) {
		priority = LOG_WARNING;
	}

	return priority;
}

/* Implementation of vfs_ops.  Pass everything on to the default
   operation but log event first. */

static int audit_connect(vfs_handle_struct *handle, const char *svc, const char *user)
{
	int result;

	result = SMB_VFS_NEXT_CONNECT(handle, svc, user);
	if (result < 0) {
		return result;
	}

	openlog("smbd_audit", LOG_PID, audit_syslog_facility(handle));

	syslog(audit_syslog_priority(handle), "connect to service %s by user %s\n", 
	       svc, user);

	return 0;
}

static void audit_disconnect(vfs_handle_struct *handle)
{
	syslog(audit_syslog_priority(handle), "disconnected\n");
	SMB_VFS_NEXT_DISCONNECT(handle);

	return;
}

static DIR *audit_opendir(vfs_handle_struct *handle,
			const struct smb_filename *smb_fname,
			const char *mask,
			uint32_t attr)
{
	DIR *result;
	
	result = SMB_VFS_NEXT_OPENDIR(handle, smb_fname, mask, attr);

	syslog(audit_syslog_priority(handle), "opendir %s %s%s\n",
	       smb_fname->base_name,
	       (result == NULL) ? "failed: " : "",
	       (result == NULL) ? strerror(errno) : "");

	return result;
}

static int audit_mkdir(vfs_handle_struct *handle,
		const struct smb_filename *smb_fname,
		mode_t mode)
{
	int result;
	
	result = SMB_VFS_NEXT_MKDIR(handle, smb_fname, mode);
	
	syslog(audit_syslog_priority(handle), "mkdir %s %s%s\n", 
	       smb_fname->base_name,
	       (result < 0) ? "failed: " : "",
	       (result < 0) ? strerror(errno) : "");

	return result;
}

static int audit_rmdir(vfs_handle_struct *handle,
		const struct smb_filename *smb_fname)
{
	int result;

	result = SMB_VFS_NEXT_RMDIR(handle, smb_fname);

	syslog(audit_syslog_priority(handle), "rmdir %s %s%s\n", 
	       smb_fname->base_name,
	       (result < 0) ? "failed: " : "",
	       (result < 0) ? strerror(errno) : "");

	return result;
}

static int audit_open(vfs_handle_struct *handle,
		      struct smb_filename *smb_fname, files_struct *fsp,
		      int flags, mode_t mode)
{
	int result;

	result = SMB_VFS_NEXT_OPEN(handle, smb_fname, fsp, flags, mode);

	syslog(audit_syslog_priority(handle), "open %s (fd %d) %s%s%s\n", 
	       smb_fname->base_name, result,
	       ((flags & O_WRONLY) || (flags & O_RDWR)) ? "for writing " : "", 
	       (result < 0) ? "failed: " : "",
	       (result < 0) ? strerror(errno) : "");

	return result;
}

static int audit_close(vfs_handle_struct *handle, files_struct *fsp)
{
	int result;

	result = SMB_VFS_NEXT_CLOSE(handle, fsp);

	syslog(audit_syslog_priority(handle), "close fd %d %s%s\n",
	       fsp->fh->fd,
	       (result < 0) ? "failed: " : "",
	       (result < 0) ? strerror(errno) : "");

	return result;
}

static int audit_rename(vfs_handle_struct *handle,
			const struct smb_filename *smb_fname_src,
			const struct smb_filename *smb_fname_dst)
{
	int result;

	result = SMB_VFS_NEXT_RENAME(handle, smb_fname_src, smb_fname_dst);

	syslog(audit_syslog_priority(handle), "rename %s -> %s %s%s\n",
	       smb_fname_src->base_name,
	       smb_fname_dst->base_name,
	       (result < 0) ? "failed: " : "",
	       (result < 0) ? strerror(errno) : "");

	return result;    
}

static int audit_unlink(vfs_handle_struct *handle,
			const struct smb_filename *smb_fname)
{
	int result;

	result = SMB_VFS_NEXT_UNLINK(handle, smb_fname);

	syslog(audit_syslog_priority(handle), "unlink %s %s%s\n",
	       smb_fname->base_name,
	       (result < 0) ? "failed: " : "",
	       (result < 0) ? strerror(errno) : "");

	return result;
}

static int audit_chmod(vfs_handle_struct *handle,
			const struct smb_filename *smb_fname,
			mode_t mode)
{
	int result;

	result = SMB_VFS_NEXT_CHMOD(handle, smb_fname, mode);

	syslog(audit_syslog_priority(handle), "chmod %s mode 0x%x %s%s\n",
	       smb_fname->base_name, mode,
	       (result < 0) ? "failed: " : "",
	       (result < 0) ? strerror(errno) : "");

	return result;
}

static int audit_chmod_acl(vfs_handle_struct *handle,
			const struct smb_filename *smb_fname,
			mode_t mode)
{
	int result;

	result = SMB_VFS_NEXT_CHMOD_ACL(handle, smb_fname, mode);

	syslog(audit_syslog_priority(handle), "chmod_acl %s mode 0x%x %s%s\n",
	       smb_fname->base_name, mode,
	       (result < 0) ? "failed: " : "",
	       (result < 0) ? strerror(errno) : "");

	return result;
}

static int audit_fchmod(vfs_handle_struct *handle, files_struct *fsp, mode_t mode)
{
	int result;

	result = SMB_VFS_NEXT_FCHMOD(handle, fsp, mode);

	syslog(audit_syslog_priority(handle), "fchmod %s mode 0x%x %s%s\n",
	       fsp->fsp_name->base_name, mode,
	       (result < 0) ? "failed: " : "",
	       (result < 0) ? strerror(errno) : "");

	return result;
}

static int audit_fchmod_acl(vfs_handle_struct *handle, files_struct *fsp, mode_t mode)
{
	int result;

	result = SMB_VFS_NEXT_FCHMOD_ACL(handle, fsp, mode);

	syslog(audit_syslog_priority(handle), "fchmod_acl %s mode 0x%x %s%s\n",
	       fsp->fsp_name->base_name, mode,
	       (result < 0) ? "failed: " : "",
	       (result < 0) ? strerror(errno) : "");

	return result;
}

static struct vfs_fn_pointers vfs_audit_fns = {
	.connect_fn = audit_connect,
	.disconnect_fn = audit_disconnect,
	.opendir_fn = audit_opendir,
	.mkdir_fn = audit_mkdir,
	.rmdir_fn = audit_rmdir,
	.open_fn = audit_open,
	.close_fn = audit_close,
	.rename_fn = audit_rename,
	.unlink_fn = audit_unlink,
	.chmod_fn = audit_chmod,
	.fchmod_fn = audit_fchmod,
	.chmod_acl_fn = audit_chmod_acl,
	.fchmod_acl_fn = audit_fchmod_acl
};

static_decl_vfs;
NTSTATUS vfs_audit_init(TALLOC_CTX *ctx)
{
	return smb_register_vfs(SMB_VFS_INTERFACE_VERSION, "audit",
				&vfs_audit_fns);
}
